### PR TITLE
add auto scaling group when creating an instance

### LINKED ISSUE #1137: add aws auto scaling groups
aws equivalent of https://github.com/nanovms/ops/issues/1135

### PR DESCRIPTION
**Pull request by olegpono**
Fixed #1137

### CODE REVIEW

File: types/config.go [hunk 1]
```diff
@@ -165,6 +165,9 @@ type ProviderConfig struct {
 
 	// AwsIPN is aws instance profile name
 	AwsIPN string
+
+	// AwsASG is aws auto scaling group
+	AwsASG string
```
Comment (eyberg) [inline types/config.go:170]: dupe?

File: types/config.go [hunk 2]
```diff
@@ -206,6 +209,12 @@ type RunConfig struct {
 	// InstanceGroup
 	InstanceGroup string
 
+	// AwsIPN is Aws instance profile name
```
Comment (eyberg) [inline types/config.go:212]: we'd like these to be cloud agnostic as much as possible so should drop the 'aws' prefix

File: types/config.go [hunk 3]
```diff
@@ -206,6 +209,12 @@ type RunConfig struct {
...
+	AwsIPN string
+
+	// AwsASG is Aws auto scaling group
+	AwsASG string
```
Comment (eyberg) [inline types/config.go:216]: I think InstanceGroup is the equivalent being used on GCP so we should use that instead

File: aws/aws_instance.go [hunk 4]
```diff
@@ -284,9 +285,47 @@ func (p *AWS) CreateInstance(ctx *lepton.Context) error {
...
+			instanceNIS.Ipv6Addresses = []*ec2.InstanceIpv6Address{addie}
 		} else {
-			instanceInput.SetIpv6AddressCount(1)
+			instanceNIS.SetIpv6AddressCount(1)
+		}
+	}
+
+	instanceInput.NetworkInterfaces = []*ec2.InstanceNetworkInterfaceSpecification{instanceNIS}
+
+	if cloudConfig.AwsIPN != "" || runConfig.AwsIPN != "" {
```
Comment (eyberg) [inline aws/aws_instance.go:296]: is this actually required for scaling groups? I don't think so?

File: aws/aws_instance.go [hunk 5]
```diff
@@ -253,29 +254,29 @@ func (p *AWS) CreateInstance(ctx *lepton.Context) error {
 	tags, tagInstanceName := buildAwsTags(cloudConfig.Tags, ctx.Config().RunConfig.InstanceName)
 	tags = append(tags, &ec2.Tag{Key: aws.String("image"), Value: &imgName})
 
+	instanceNIS := &ec2.InstanceNetworkInterfaceSpecification{
+		AssociatePublicIpAddress: aws.Bool(true),
```
Comment (eyberg) [inline aws/aws_instance.go:258]: this shouldn't be hard-coded

**Review by eyberg — CHANGES_REQUESTED** [review]
i'd make all those config names cloud-agnostic - for instance we already have' instance group' for gcp

File: aws/aws_instance.go (hunk 4 above)
Comment (olegpono) [reply aws/aws_instance.go]: Aws IntanceProfileName is optional

File: cmd/cmd_instance.go [hunk 6]
```diff
@@ -48,6 +48,8 @@ func instanceCreateCommand() *cobra.Command {
 	PersistCreateInstanceFlags(cmdInstanceCreate.PersistentFlags())
 	cmdInstanceCreate.PersistentFlags().StringP("instance-name", "i", "", "instance name (overrides default instance name format)")
 	cmdInstanceCreate.PersistentFlags().StringP("instance-group", "", "", "instance group")
+	cmdInstanceCreate.PersistentFlags().StringP("aws-ipn", "", "", "AWS instance profile name")
+	cmdInstanceCreate.PersistentFlags().StringP("aws-asg", "", "", "AWS auto scaling group")
```
Comment (eyberg) [inline cmd/cmd_instance.go:52]: this should be the same as instance group right?

File: cmd/cmd_instance.go [hunk 7]
```diff
@@ -48,6 +48,8 @@ func instanceCreateCommand() *cobra.Command {
 	PersistCreateInstanceFlags(cmdInstanceCreate.PersistentFlags())
 	cmdInstanceCreate.PersistentFlags().StringP("instance-name", "i", "", "instance name (overrides default instance name format)")
 	cmdInstanceCreate.PersistentFlags().StringP("instance-group", "", "", "instance group")
+	cmdInstanceCreate.PersistentFlags().StringP("aws-ipn", "", "", "AWS instance profile name")
```
Comment (eyberg) [inline cmd/cmd_instance.go:51]: if we have the equivalent of the GCP version of this we should use this and if not just make it agnostic so it's not aws specific

**Review by eyberg — APPROVED** [review]

**Pull request merged**